### PR TITLE
fix: streaming cat over http api

### LINKED ISSUE #1755: /api/v0/cat buffers output
https://github.com/ipfs/js-ipfs/blob/master/src/http/api/resources/files-regular.js#L82

### PR DESCRIPTION
**Pull request by alanshaw**
`/api/v0/cat` calls `ipfs.cat`, but `ipfs.cat` returns a buffer of file output. This changes `/api/v0/cat` to actually stream output by calling `ipfs.catPullStream` instead.

It uses `pull-pushable` in order to catch an initial error in the stream before it starts flowing and instead return a plain JSON response with an appropriate HTTP status.

This isn't ideal as it means there's no backpressure - if the consumer doesn't consume fast enough then data will start to get buffered into memory. However this is significantly better than buffering _everything_ into memory before replying.

This reduces memory usage and time to first byte significantly. e.g.

```sh
$ jsipfs object stat QmTtBrkubV9M914kgYrjTDCWTkN4TjnwYDJYb3dwK7ErXR
NumLinks: 22
BlockSize: 1110
LinksSize: 992
DataSize: 118
CumulativeSize: 965374843 # i.e. ~965MB

# I modified jsipfs cat to print out the time it takes before it receives the first byte:
# (note we are talking to the daemon here so are going over the HTTP API)

# Before:
$ jsipfs cat QmTtBrkubV9M914kgYrjTDCWTkN4TjnwYDJYb3dwK7ErXR
time to first chunk: 2961.353ms

# After:
$ jsipfs cat QmTtBrkubV9M914kgYrjTDCWTkN4TjnwYDJYb3dwK7ErXR
time to first chunk: 127.899ms
```

resolves #1755

### CODE REVIEW

**Comment by daviddias** [comment]
🤯🤯🤯🤯🤯🤯

File: src/core/components/files-regular.js [hunk 1]
```diff
@@ -232,21 +232,24 @@ module.exports = function (self) {
...
+
+        if (!files.length) {
           return d.abort(new Error('No such file'))
         }
 
         const file = files[0]
-        const content = file.content
-        if (!content && file.type === 'dir') {
+
+        if (!file.content && file.type === 'dir') {
```
Comment (achingbrain) [inline src/core/components/files-regular.js]: These conditions might be simpler as:

```suggestion
        if (file.type !== 'file') {
          return d.abort(new Error('this dag node was not a file')
        }
        
        if (!file.content) {
          return d.abort(new Error('this dag node had no content')
        }
```

Though they could also be added to the `filterFile` function.

Comment (achingbrain) [reply src/core/components/files-regular.js]: Fair enough.  I guess once all the errors have codes we can stop depending on message content.

If `file.type === 'file'` then `file.content` should always have a value.  It's weird that there is a check for that in the first place 🤷‍♂️.

**Comment by alanshaw** [comment]
Merging as test failures seem to be because of unrelated IPNS tests.

**Pull request merged**